sequencer/src/bin/espresso-dev-node.rs

```diff
@@ -148,52 +162,130 @@ async fn main() -> anyhow::Result<()> {
-            .phrase(cli_params.mnemonic.as_str())
-            .index(cli_params.account_index)
-            .expect("error building wallet")
-            .build()
-            .expect("error opening wallet")
-            .with_chain_id(chain_id)
-            .signer()
-            .clone(),
+        l1_provider: url.clone(),
+        light_client_address,
+        eth_signing_key: wallet.signer().clone(),
         sequencer_url: "http://localhost".parse().unwrap(), // This should not be used in dev-node
-        port: None,
+        port: Some(prover_port),
         stake_table_capacity: STAKE_TABLE_CAPACITY_FOR_TEST as usize,
     };
 
-    loop {
-        if let Err(err) = sync_state(&st, proving_key.clone(), &relay_server_client, &config).await
-        {
-            tracing::error!("Cannot sync the light client state, will retry: {}", err);
-            sleep(retry_interval).await;
-        } else {
-            tracing::info!("Sleeping for {:?}", update_interval);
-            sleep(update_interval).await;
+    spawn(run_prover_service_with_stake_table(
+        config,
+        SEQUENCER_VERSION,
+        Arc::new(st),
+    ));
+
+    let dev_info = DevInfo {
+        builder_url: network.cfg.hotshot_config().builder_urls[0].clone(),
+        prover_port,
+        l1_url: url,
+        light_client_address,
+    };
+
+    let mock_contract =
+        LightClientMock::new(light_client_address, Arc::new(provider.with_signer(wallet)));
+
+    run_dev_node_server(
+        cli_params.dev_node_port,
+        mock_contract,
+        dev_info,
+        SEQUENCER_VERSION,
+    )
+    .await?;
+
+    Ok(())
+}
+
+async fn run_dev_node_server<Ver: StaticVersionType + 'static, S: Signer + Clone + 'static>(
+    port: u16,
+    mock_contract: LightClientMock<SignerMiddleware<Provider<Http>, S>>,
+    dev_info: DevInfo,
+    bind_version: Ver,
+) -> anyhow::Result<()> {
+    let mut app = tide_disco::App::<(), ServerError>::with_state(());
+    let toml =
+        toml::from_str::<toml::value::Value>(include_str!("../../api/espresso_dev_node.toml"))
+            .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+    let mut api = Api::<(), ServerError, Ver>::new(toml)
+        .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+    api.get("devinfo", move |_, _| {
+        let info = dev_info.clone();
+        async move { Ok(info.clone()) }.boxed()
+    })
+    .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+    let contract = mock_contract.clone();
+    api.get("freeze", move |req, _| {
+        let contract = contract.clone();
+        async move {
+            let height: u64 = req
+                .integer_param("height")
+                .map_err(ServerError::from_request_error)?;
+            contract
+                .set_hot_shot_down_since(U256::from(height))
+                .send()
+                .await
+                .map_err(|err| ServerError::catch_all(StatusCode::FORBIDDEN, err.to_string()))?;
```

Comment:
Why is the status code `FORBIDDEN` here?

Reply:
Maybe `BAD_REQUEST` is better?

Reply:
maybe `INTERNAL_SERVER_ERROR` ?

Reply:
e92c71737d74e7e3db24e017de9bf791f5bfaaac